crates/wit/src/commands/pull.rs

```diff
@@ -0,0 +1,273 @@
+use std::{
+    collections::{BTreeSet, HashSet},
+    io::Write,
+    path::PathBuf,
+};
+
+use anyhow::{bail, Context, Result};
+use clap::Args;
+
+use cargo_component_core::{command::CommonOptions, VersionedPackageName};
+use futures::TryStreamExt;
+use tokio_util::io::{StreamReader, SyncIoBridge};
+use warg_loader::{ClientConfig, Release};
+use wit_component::DecodedWasm;
+use wit_parser::{PackageId, PackageName, Resolve, UnresolvedPackage};
+
+/// Pull WIT package(s) to a local "deps" directory.
+#[derive(Args)]
+#[clap(disable_version_flag = true)]
+pub struct PullCommand {
+    /// The common command options.
+    #[clap(flatten)]
+    pub common: CommonOptions,
+
+    /// Use the specified registry name when pulling the package(s).
+    #[clap(long, value_name = "REGISTRY")]
+    pub registry: Option<String>,
+
+    /// Update the specified directory WIT "root" directory. Dependencies will
+    /// be written to e.g. "<wit-dir>/deps/<namespace>.<package>.wit".
+    #[clap(long, value_name = "WIT_DIR", default_value = "wit")]
+    pub wit_dir: PathBuf,
+
+    /// Create "<wit-dir>" and "<wit-dir>/deps" directories if needed.
+    #[clap(long)]
+    pub create_dirs: bool,
+
+    /// Pull the packages specified. If empty, the list of packages to pull
+    /// will be inferred from missing dependencies.
+    #[clap(value_name = "PACKAGE")]
+    pub packages: Vec<VersionedPackageName>,
+}
+
+impl PullCommand {
+    /// Executes the command.
+    pub async fn exec(self) -> Result<()> {
+        log::debug!("executing pull command");
+
+        let terminal = self.common.new_terminal();
+
+        let (mut existing_pkgs, deps) = self.parse_packages_state()?;
+        log::debug!("Packages state: pkgs={existing_pkgs:?} deps={deps:?}");
+
+        let packages = if self.packages.is_empty() {
+            // No packages specified; look for missing dependencies
+            deps.into_iter()
+                .filter(|pkg| !existing_pkgs.contains(pkg))
+                .map(|pkg| VersionedPackageName {
+                    name: wit_to_warg_package_name(&pkg),
+                    version: pkg.version.map(|ver| ver.to_string().parse().unwrap()),
+                })
+                .collect::<Vec<_>>()
+        } else {
+            // Remove existing packages from given list
+            self.packages
+                .iter()
+                .filter(|pkg| {
+                    !existing_pkgs
+                        .iter()
+                        .any(|existing| version_satisfied(pkg, existing))
+                })
+                .cloned()
+                .collect()
+        };
+
+        if packages.is_empty() {
+            terminal.status("Finished", "no missing packages; nothing to do")?;
+            return Ok(());
+        }
+
+        let mut client = {
+            let mut config = ClientConfig::default();
+            config.namespace_registry("wasi", "bytecodealliance.org");
```

Comment:
Now I'm not sure what the best UX is here. If you specify `--registry` you definitely expect at least some package to come from there, but then should it fall back to the config file / defaults otherwise? :thinking:

Reply:
I'm going to punt a bit and do the thing that is currently easy, which is for `--registry` to only set the _default_ registry which will still be overridden by other configured namespace mappings. I don't think this is ideal but should work most of the time. a0e0e51